File: src/main/java/de/tum/www1/orion/ui/browser/BrowserWebView.java
package de.tum.www1.orion.ui.browser;

import com.intellij.ide.DataManager;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import de.tum.www1.orion.bridge.core.ArtemisCoreUpcallBridge;
import de.tum.www1.orion.bridge.downcall.ArtemisJavascriptDowncallBridge;
import de.tum.www1.orion.bridge.instructor.ArtemisInstructorUpcallBridge;
import de.tum.www1.orion.bridge.test.ArtemisTestResultReporter;
import de.tum.www1.orion.ui.OrionRouter;
import de.tum.www1.orion.ui.util.UrlAccessForbiddenWarning;
import de.tum.www1.orion.util.OrionSettingsProvider;
import de.tum.www1.orion.util.registry.OrionInstructorExerciseRegistry;
import javafx.application.Platform;
import javafx.concurrent.Worker;
import javafx.embed.swing.JFXPanel;
import javafx.scene.Scene;
import javafx.scene.layout.BorderPane;
import javafx.scene.web.WebEngine;
import javafx.scene.web.WebView;
import netscape.javascript.JSObject;

import javax.swing.*;
import java.util.Objects;
import java.util.ResourceBundle;

public class BrowserWebView {
    private WebView browser;
    private WebEngine engine;
    private JFXPanel browserPanel;
    private Project project;

    /**
     * Inits the actual browser panel. We use a JFXPanel in a {@link WebView} gets initialized. This web view only
     * displays the ArTEMiS Angular webapp containing a few adaptions, so that we only show the most important information
     * in the IDE.
     */
    public void init() {
        browserPanel = new JFXPanel();
        Platform.setImplicitExit(false);
        Platform.runLater(() -> {
            browser = new WebView();
            engine = browser.getEngine();
            final var version = ResourceBundle.getBundle("de.tum.www1.orion.Orion").getString("version");
            engine.setUserAgent(ServiceManager.getService(OrionSettingsProvider.class).getSetting(OrionSettingsProvider.KEYS.USER_AGENT) + " Orion/" + version);
            project = Objects.requireNonNull(DataManager.getInstance().getDataContext(browserPanel).getData(CommonDataKeys.PROJECT));

            final OrionRouter orionRouter = ServiceManager.getService(project, OrionRouter.class);
            final String route = orionRouter.routeForCurrentExercise();
            engine.load(Objects.requireNonNullElseGet(route, orionRouter::defaultRoute));

            injectJSBridge();
            preventAccessToExternalWebpages();
        });
    }

    private void injectJSBridge() {
        engine.getLoadWorker().stateProperty().addListener((observableValue, state, t1) -> {
            if (state == Worker.State.SUCCEEDED || t1 == Worker.State.SUCCEEDED) {
                final var window = (JSObject) engine.executeScript("window");
                final var coreBridge = ServiceManager.getService(project, ArtemisCoreUpcallBridge.class);
                coreBridge.attachTo(window, "orionCoreConnector");

                final var registry = ServiceManager.getService(project, OrionInstructorExerciseRegistry.class);
                if (registry.isArtemisExercise()) {
                    final var testResultBridge = ServiceManager.getService(project, ArtemisTestResultReporter.class);
                    testResultBridge.attachTo(window, "orionTestResultsConnector");

                    if (registry.isOpenedAsInstructor()) {
                        final var instructorBridge = ServiceManager.getService(project, ArtemisInstructorUpcallBridge.class);
                        instructorBridge.attachTo(window, "orionInstructorConnector");
                    }
                }

                ServiceManager.getService(project, ArtemisJavascriptDowncallBridge.class).artemisLoadedWith(engine);
            }
        });
    }

    private void preventAccessToExternalWebpages() {
        this.engine.locationProperty().addListener(((observable, oldValue, newValue) -> {
            if (!newValue.startsWith(ServiceManager.getService(project, OrionRouter.class).defaultRoute())) {
                engine.load(oldValue);
                ApplicationManager.getApplication().invokeLater(() -> new UrlAccessForbiddenWarning(project).show());
            }
        }));
    }

    /**
     * Get the browser panel in which the ArTEMiS webapp is displayed.
     *
     * @return The {@link JFXPanel} containing the {@link WebView} with ArTEMiS
     */
    public JComponent getBrowser() {
        Platform.runLater(() -> {
            final BorderPane borderPane = new BorderPane();
            borderPane.setCenter(browser);
            final Scene scene = new Scene(borderPane);
            browserPanel.setScene(scene);
        });

        return browserPanel;
    }
}

File: src/main/java/de/tum/www1/orion/ui/browser/BrowserWebView.kt
package de.tum.www1.orion.ui.browser

import com.intellij.openapi.components.ServiceManager
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import com.intellij.ui.jcef.JBCefApp
import com.intellij.ui.jcef.JBCefBrowser
import com.intellij.ui.jcef.JBCefClient
import com.intellij.ui.jcef.JBCefJSQuery
import com.intellij.util.messages.Topic
import de.tum.www1.orion.connector.ide.build.OrionBuildConnector
import de.tum.www1.orion.connector.ide.exercise.OrionExerciseConnector
import de.tum.www1.orion.connector.ide.shared.OrionSharedUtilConnector
import de.tum.www1.orion.connector.ide.vcs.OrionVCSConnector
import de.tum.www1.orion.settings.OrionSettingsProvider
import de.tum.www1.orion.ui.OrionRouter
import org.cef.CefApp
import org.cef.CefSettings
import org.cef.browser.CefBrowser
import org.cef.browser.CefFrame
import org.cef.handler.CefLoadHandlerAdapter
import java.util.*

class BrowserWebView(val project: Project) {
    val jbCefBrowser: JBCefBrowser
    val jsQuery: JBCefJSQuery
    val client: JBCefClient

    /**
     * Inits the actual browser panel. We use a JFXPanel in a [WebView] gets initialized. This web view only
     * displays the ArTEMiS Angular webapp containing a few adaptions, so that we only show the most important information
     * in the IDE.
     */
    init {
        //Dumb check but must be called to perform initialization
        if (!JBCefApp.isSupported()) {
            // Fallback to an alternative browser-less solution
        }
        val version = ResourceBundle.getBundle("de.tum.www1.orion.Orion").getString("version")
        val userAgent=ServiceManager.getService(OrionSettingsProvider::class.java).getSetting(OrionSettingsProvider.KEYS.USER_AGENT) + " Orion/" + version
        val route = project.service<OrionRouter>().routeForCurrentExercise()
        //Since JBCef wrapper doesn't support setting user-agent, we need to use reflection to access private properties.
        val jbCefAppInstance= JBCefApp.getInstance()
        val privateCefApp=jbCefAppInstance.getPrivateProperty<CefApp>("myCefApp")
        val privateCefSettings=privateCefApp.getPrivateProperty<CefSettings>("settings_")
        //The user agent needes to set before the call to createClient and can not be changed after
        privateCefSettings.user_agent=userAgent
        client=jbCefAppInstance.createClient()
        jbCefBrowser=JBCefBrowser(client, route)
        client.addLoadHandler(object : CefLoadHandlerAdapter() {
            override fun onLoadEnd(browser: CefBrowser?, frame: CefFrame?, httpStatusCode: Int) {
                if(browser==null)
                    throw NullPointerException("Browser in onLoadEnd should not be null")
                else project.messageBus.syncPublisher(OrionBrowserNotifier.ORION_BROWSER_TOPIC).artemisLoadedWith(browser)
            }
        }, jbCefBrowser.cefBrowser)
        jsQuery = JBCefJSQuery.create(jbCefBrowser)
        injectJSBridge()
    }

    private fun injectJSBridge() {
        OrionSharedUtilConnector(this).initializeHandlers()
        OrionExerciseConnector(this).initializeHandlers()
        OrionBuildConnector(this).initializeHandlers()
        OrionVCSConnector(this).initializeHandlers()
    }
}

inline fun <reified E> Any.getPrivateProperty(propertyName: String): E {
    val privatePropertyField=this.javaClass.getDeclaredField(propertyName).apply {
        isAccessible=true
    }
    return privatePropertyField.get(this) as E
}

interface OrionBrowserNotifier {
    fun artemisLoadedWith(engine: CefBrowser)

    companion object {
        @JvmField
        val ORION_BROWSER_TOPIC = Topic.create("Orion Browser Init", OrionBrowserNotifier::class.java)
    }
}